code/obj/machinery/computer/alert.dm

```diff
@@ -0,0 +1,155 @@
+
+	//Oh, someone is asking us for data instead of reporting a thing.
+	if((signal.data["command"] == "report_alerts") && signal.data["sender"])
+		src.generate_signal_report(signal)
+		return
+
+	if(signal.data["command"] != "update_alert")
+		return
+
+	var/zone = signal.data["zone"]
+	var/severity = signal.data["alert"]
+	var/kind = signal.data["type"]
+
+	if(!zone || !severity || !kind) return
+
+	var/datum/alert/new_alert = new /datum/alert
+	new_alert.zone = zone
+	new_alert.severity = severity
+	new_alert.kind = kind
+
+	src.update_alerts(new_alert)
+
+///Generate the report used for the EngieAlerter PDA app
+/obj/machinery/computer/general_alert/proc/generate_signal_report(datum/signal/signal)
+	var/datum/signal/newsignal = get_free_signal()
+	newsignal.data["address_1"] = signal.data["sender"]
+	newsignal.data["command"] = "reply_alerts"
+	var/list/priority = list()
+	var/list/minor = list()
+	for (var/datum/alert/alert in src.alerts)
+		switch(alert.severity)
+			if(ALERT_SEVERITY_PRIORITY)
+				priority += alert.zone
+			if(ALERT_SEVERITY_MINOR)
+				minor += alert.zone
+	if (length(priority))
+		newsignal.data["severe_list"] = jointext(priority, ";")
+	if (length(minor))
+		newsignal.data["minor_list"] = jointext(minor, ";")
+	SEND_SIGNAL(src, COMSIG_MOVABLE_POST_RADIO_PACKET, newsignal, null, "respond")
+
+///Update the ale
+/obj/machinery/computer/general_alert/proc/update_alerts(datum/alert/alert_update)
+	for (var/datum/alert/existing_alert in src.alerts)
+		if (existing_alert.zone == alert_update.zone)
+			if(existing_alert.kind == alert_update.kind)
+				src.alerts -= existing_alert
+				break
+	if(alert_update.severity == ALERT_SEVERITY_RESET)
+		return
+	src.alerts += alert_update
+	src.update_alert_icon()
+
+/obj/machinery/computer/general_alert/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.) return
+
+	var/removing_zone = params["area_ckey"]
+	for (var/datum/alert/alert in src.alerts)
+		if(ckey(alert.zone) == removing_zone)
+			switch(action) {
+				if("clear_atmos")
+					if(alert.kind == ALERT_KIND_ATMOS)
+						src.alerts -= alert
+						break
+				if("clear_fire")
+					if(alert.kind == ALERT_KIND_FIRE)
+						src.alerts -= alert
+						break
+				if("clear_power")
+					if(alert.kind == ALERT_KIND_POWER)
+						src.alerts -= alert
+						break
+			}
+	src.update_alert_icon()
+
+/obj/machinery/computer/general_alert/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "AlertComputer")
+		ui.open()
+
+/obj/machinery/computer/general_alert/ui_data(mob/user)
+	. = ..()
+	.["alerts"] = list()
+	for (var/datum/alert/alert in src.alerts)
+		var/area_ckey = ckey(alert.zone)
+		if(!(area_ckey in .["alerts"]))
+			.["alerts"][area_ckey] = list()
+		.["alerts"][area_ckey]["area_ckey"] = area_ckey
+		.["alerts"][area_ckey]["area_name"] = alert.zone
+		.["alerts"][area_ckey][alert.kind] = alert.severity
```

Comment:
Your response to the previous comment re: UI data structure had me come look here. This line in particular sets of a code smell alert (geddit? Because of what you're working on).

If I'm reading correctly, this only gives one alert per area, as if multiple entries in `src.alerts` have the same area the latest one overwrites the first one.

I'm not opposed to grouping by area, but subsequent alerts should add a new list item, not overwrite the fields on the previous one in the same area, right?

Reply:
updated the alerts to have one datum per area ("zone"), which made this logic a lot clearer